### PR TITLE
Add unit tests for org.jkiss.utils.BeanUtils

### PR DESCRIPTION
**Pull request by EricHetti**
I've analysed your codebase and noticed that `org.jkiss.utils.BeanUtils` is not fully tested.
I've written some tests for the methods in this class with the help of [Diffblue Cover](https://www.diffblue.com/opensource).

Hopefully, these tests will help you detect any regressions caused by future code changes. If you would find it useful to have additional tests written for this repository, I would be more than happy to look at other classes that you consider important.

### CODE REVIEW

File: bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+        Assert.assertNull(BeanUtils.getGetMethod(String.class, "length"));
+        Assert.assertNull(
+                BeanUtils.getGetMethod(String.class, "length", true));
+        Assert.assertNull(
+                BeanUtils.getGetMethod(String.class, "length", false));
+    }
+
+    @Test
+    public void testMethodNameToPropertyName() {
+        Assert.assertNull(BeanUtils.methodNameToPropertyName(""));
```
Comment (petersear) [inline bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java:100]: Add blank line between different assert types.

File: bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    @Test
+    public void testGetGetMethod() {
+        Assert.assertNull(BeanUtils.getGetMethod(String.class, "size"));
+        Assert.assertNull(BeanUtils.getGetMethod(String.class, "length"));
+        Assert.assertNull(
+                BeanUtils.getGetMethod(String.class, "length", true));
+        Assert.assertNull(
+                BeanUtils.getGetMethod(String.class, "length", false));
+    }
+
```
Comment (petersear) [inline bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java:85]: There is no test for `propertyNameToMethodName()`. Not sure if this was intentional as it looks coverable and other methods seems to be present.

File: bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+                PowerMockito.mock(ParameterizedType.class);
+        PowerMockito.when(parameterizedType.getActualTypeArguments())
+                .thenReturn(new Type[]{wildcardType});
+
+        Assert.assertEquals(String.class,
+                BeanUtils.getCollectionType(parameterizedType));
+    }
+
+    @Test
+    public void testReadObjectProperty()
```
Comment (petersear) [inline bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java:218]: Can you assert any non-null values here?

File: bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+                BeanUtils.getDefaultPrimitiveValue(Character.TYPE));
+        Assert.assertEquals(false,
+                BeanUtils.getDefaultPrimitiveValue(Boolean.TYPE));
+
+        thrown.expect(IllegalArgumentException.class);
+        BeanUtils.getDefaultPrimitiveValue(String.class);
+    }
+
+    @Test
+    public void testIsNumericType() {
```
Comment (petersear) [inline bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java:261]: Add a false case

**Review by petersear — CHANGES_REQUESTED** [review]

File: bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java (hunk 1 above)
Comment (Braavos6) [reply bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java:100]: Done

File: bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java (hunk 2 above)
Comment (Braavos6) [reply bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java:85]: Ooops.. Done

File: bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java (hunk 3 above)
Comment (Braavos6) [reply bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java:218]: Done

File: bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java (hunk 4 above)
Comment (Braavos6) [reply bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java:261]: Done

File: bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+        Assert.assertNull(BeanUtils.getGetMethod(String.class, "size"));
+        Assert.assertNull(BeanUtils.getGetMethod(String.class, "length"));
+        Assert.assertNull(
+                BeanUtils.getGetMethod(String.class, "length", true));
+        Assert.assertNull(
+                BeanUtils.getGetMethod(String.class, "length", false));
+    }
+
+    @Test
+    public void testPropertyNameToMethodName() {
```
Comment (petersear) [inline bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java:87]: Looks good except that `length` is misspelt in each case.

**Review by petersear — CHANGES_REQUESTED** [review]

File: bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java (hunk 5 above)
Comment (Braavos6) [reply bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java:87]: Oops Done

**Review by petersear — APPROVED** [review]

**Comment by petersear** [comment]
This PR compares changes with the add-diffblue-yml branch, the top commit of which is to for adding the diffblue.yml file, so don't forget to remove that commit before submitting the final PR.

File: bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+        Assert.assertEquals(String.class,
+                BeanUtils.getCollectionType(parameterizedType));
+
+        Assert.assertNull(BeanUtils.getCollectionType(null));
+    }
+
+    @Test
+    public void testGetCollectionType2() {
+        WildcardType wildcardType = PowerMockito.mock(WildcardType.class);
```
Comment (JohnLBergqvist) [inline bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java]: I'm not sure `wildcardType` is the should be mocked if that's what you're testing...

File: bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+        PowerMockito.when(wildcardType.getUpperBounds())
+                .thenReturn(new Type[]{String.class, Integer.class});
+
+        ParameterizedType parameterizedType =
```
Comment (JohnLBergqvist) [inline bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java]: Same for `parameterisedType`

File: bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+        PowerMockito.when(parameterizedType.getActualTypeArguments())
+                .thenReturn(new Type[]{wildcardType});
+
+        Assert.assertEquals(String.class,
+                BeanUtils.getCollectionType(parameterizedType));
+    }
+
+    @Test
+    public void testGetCollectionType3() {
+        WildcardType wildcardType = PowerMockito.mock(WildcardType.class);
```
Comment (JohnLBergqvist) [inline bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java]: ditto.

**Review by JohnLBergqvist — CHANGES_REQUESTED** [review]

File: bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java (hunk 6 above)
Comment (Braavos6) [reply bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java]: As both `wildcardType` and `ParameterizedType` are interfaces they will have to be implemented. 

So would you like me to do this instead?
```
    @Test
    public void testGetCollectionType2() {
        WildcardType wildcardType = PowerMockito.mock(WildcardType.class);
        PowerMockito.when(wildcardType.getUpperBounds())
                .thenReturn(new Type[]{String.class, Integer.class});

        ParameterizedType parameterizedType = new ParameterizedType() {
            @Override
            public Type[] getActualTypeArguments() {
                return new Type[]{wildcardType};
            }

            @Override
            public Type getRawType() {
                return null;
            }

            @Override
            public Type getOwnerType() {
                return null;
            }
        };

        Assert.assertEquals(String.class,
                BeanUtils.getCollectionType(parameterizedType));
    }
```

Comment (JohnLBergqvist) [reply bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java]: If that still works, yes.

Comment (Braavos6) [reply bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java]: Done

File: bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java (hunk 7 above)
Comment (Braavos6) [reply bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java]: Done

File: bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java (hunk 8 above)
Comment (Braavos6) [reply bundles/org.jkiss.utils/tests/org/jkiss/utils/BeanUtilsTest.java]: Done

**Review by JohnLBergqvist — APPROVED** [review]

**Pull request closed without merging**